src/ape_cache/query.py

```diff
@@ -0,0 +1,337 @@
+
+    def init_db(self):
+        if self.database_file.is_file():
+            raise QueryEngineError("Database has already been initialized")
+
+        models.Base.metadata.create_all(bind=self.engine)
+
+    def purge_db(self):
+        if not self.database_file.is_file():
+            # Add check here to show we have a file that exists
+            raise QueryEngineError("Database must be initialized")
+
+        self.database_file.unlink()
+
+    def _block_query(
+        self, query_stmt: str, query: Union[BlockQuery, ContractEventQuery]
+    ) -> Union[int, pd.DataFrame]:
+        with self.engine.connect() as conn:
+            if "COUNT" in query_stmt:
+                q = conn.execute(
+                    text(query_stmt),
+                    start_block=query.start_block,
+                    stop_block=query.stop_block,
+                    step=query.step,
+                )
+                return q.rowcount
+            elif "SELECT :" in query_stmt:
+                q = conn.execute(
+                    text(query_stmt),
+                    columns=",".join(query.columns),
+                    start_block=query.start_block,
+                    stop_block=query.stop_block,
+                    step=query.step,
+                )
+                return pd.DataFrame(columns=query.columns, data=q.fetchall())
+        raise Exception("SELECT statement improperly set.")
+
+    @singledispatchmethod
+    def table(self, query: QueryType):
+        raise QueryEngineError("Not a compatible QueryType")
+
+    @table.register
+    def block_table(self, query: BlockQuery) -> str:
+        return "blocks"
+
+    @table.register
+    def transactions_table(self, query: BlockTransactionQuery) -> str:
+        return "transactions"
+
+    @table.register
+    def contract_events_table(self, query: ContractEventQuery) -> str:
+        return "contract_events"
+
+    @singledispatchmethod
+    def column(self, query: QueryType) -> str:
+        raise QueryEngineError("Not a compatible QueryType")
+
+    @column.register
+    def block_column(self, query: BlockQuery) -> str:
+        return "number"
+
+    @column.register
+    def block_transaction_column(self, query: BlockTransactionQuery) -> str:
+        return "block_hash"
+
+    @column.register
+    def contract_event_column(self, query: ContractEventQuery) -> str:
+        return "block_number"
+
+    @singledispatchmethod
+    def cache_query(self, query: QueryType) -> str:
+        raise QueryEngineError("Not a compatible QueryType")
+
+    @cache_query.register
+    def block_cache_query(self, query: BlockQuery) -> str:
+        return "SELECT * FROM blocks WHERE number = :val"
+
+    @cache_query.register
+    def block_transaction_cache_query(self, query: BlockTransactionQuery) -> str:
+        return "SELECT * FROM transactions WHERE block_hash = :val"
+
+    @cache_query.register
+    def contract_event_cache_query(self, query: ContractEventQuery) -> str:
```

Comment:
Are we able to query off other fields besides block number, such as topics, for events?

Reply:
Can add support for this in a later PR